File: database/database-annotation-processor/src/main/java/ru/tinkoff/kora/database/annotation/processor/model/QueryParameter.java
package ru.tinkoff.kora.database.annotation.processor.model;

import ru.tinkoff.kora.database.annotation.processor.entity.DbEntity;

import javax.lang.model.element.VariableElement;
import javax.lang.model.type.TypeMirror;

public sealed interface QueryParameter {
    String name();

    TypeMirror type();

    VariableElement variable();

    record ConnectionParameter(String name, TypeMirror type, VariableElement variable) implements QueryParameter {}

    record SimpleParameter(String name, TypeMirror type, VariableElement variable) implements QueryParameter {}

    record EntityParameter(String name, DbEntity entity, VariableElement variable) implements QueryParameter {
        @Override
        public TypeMirror type() {
            return entity.typeMirror();
        }
    }

    record BatchParameter(String name, TypeMirror type, VariableElement variable, QueryParameter parameter) implements QueryParameter {
        public BatchParameter {
            if (!(parameter instanceof SimpleParameter || parameter instanceof EntityParameter)) {
                throw new IllegalStateException();
            }
        }
    }
}

File: database/database-symbol-processor/src/main/kotlin/ru/tinkoff/kora/database/symbol/processor/model/QueryParameter.kt

package ru.tinkoff.kora.database.symbol.processor.model

import com.google.devtools.ksp.symbol.KSType
import com.google.devtools.ksp.symbol.KSValueParameter

sealed interface QueryParameter {
    val name: String
    val type: KSType
    val variable: KSValueParameter

    data class ConnectionParameter(override val name: String, override val type: KSType, override val variable: KSValueParameter) : QueryParameter

    data class SimpleParameter(override val name: String, override val type: KSType, override val variable: KSValueParameter) : QueryParameter

    data class EntityParameter(override val name: String, override val type: KSType, override val variable: KSValueParameter, val entity: DbEntity) : QueryParameter

    data class BatchParameter(override val name: String, override val type: KSType, override val variable: KSValueParameter, val parameter: QueryParameter) : QueryParameter
}
